### PR TITLE
feat: `slinky` Integration

### PR DESCRIPTION
**Pull request by ajansari95**
## 1. Overview
- Integrates Slinky f
- Fixes CLI bug related to double wiring of tx commands
- Implements ABCI++ vote extensions
- Adds v4 upgrade plan with Slinky integration

## 2. Implementation Details
- Integrated `x/oracle` and `x/marketmap` modules from Skip
- Defined an aggregate function for pre-block price aggregation
- Added v4 upgrade plan and corresponding integration tests
- Resolved tx command duplication in CLI

## 3. Testing and Usage
- Verify CLI commands: `./build/quasard tx`
- Test v4 upgrade: test in `app/upgrades/v4/upgrades_test.go`

## 6. Future Work
- Explore custom data feeds in Slinky sidecar process
- Consider upstreaming Quasar-specific data feed requirements to Slinky

### CODE REVIEW

File: app/upgrades/v3/upgrades_test.go [hunk 1]
```diff
@@ -45,7 +45,8 @@ func (s *UpgradeTestSuite) TestUpgrade() {
 	ctx := s.Ctx
 	keepers := s.App.AppKeepers
 
-	modulesToRemove := []string{"qtransfer", "qoracle", "qvesting"}
+	modulesToRemove := []string{}
+	//modulesToRemove := []string{"qtransfer", "qoracle", "qvesting"}
```
Comment (lubkoll) [inline app/upgrades/v3/upgrades_test.go:49]: If this is intentional, could you remove the commented out part

Comment (ajansari95) [reply app/upgrades/v3/upgrades_test.go]: [25a3390](https://github.com/quasar-finance/quasar/pull/859/commits/25a33901594d5d867de775151e2f797be28cb9d2)

File: app/upgrades/v4/constants.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import (
+	storetypes "cosmossdk.io/store/types"
+	"github.com/quasar-finance/quasar/app/upgrades"
+	marketmaptypes "github.com/skip-mev/slinky/x/marketmap/types"
+	oracletypes "github.com/skip-mev/slinky/x/oracle/types"
+)
+
+const (
+	UpgradeName                = "v4"
+	MarketMapAuthorityMultisig = "" //TODO: add
```
Comment (magiodev) [inline app/upgrades/v4/constants.go:12]: todo??

File: app/upgrades/v4/upgrades.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+		ctx.Logger().Info(fmt.Sprintf("Migration {%s} applied", UpgradeName))
+		return vm, nil
+	}
+}
+
+func setMarketMapParams(ctx sdk.Context, marketmapKeeper *marketmapkeeper.Keeper) error {
+	marketmapParams := marketmaptypes.Params{
+		MarketAuthorities: []string{authtypes.NewModuleAddress(govtypes.ModuleName).String()}, //MarketMapAuthorityMultisig},
+		Admin:             authtypes.NewModuleAddress(govtypes.ModuleName).String(),           // TODO: decide admin
```
Comment (magiodev) [inline app/upgrades/v4/upgrades.go:62]: todo

File: app/upgrades/v3/upgrades.go [hunk 4]
```diff
@@ -27,7 +27,7 @@ func CreateUpgradeHandler(
 	keepers *keepers.AppKeepers,
 ) upgradetypes.UpgradeHandler {
 	return func(c context.Context, plan upgradetypes.Plan, fromVM module.VersionMap) (module.VersionMap, error) {
-		modulesToRemove := []string{"qtransfer", "qoracle", "qvesting"}
+		modulesToRemove := []string{}
```
Comment (arhamchordia) [inline app/upgrades/v3/upgrades.go:30]: We can completely get rid of this v3 upgrade directory here.

File: app/upgrades/v3/upgrades_test.go [hunk 5]
```diff
@@ -45,7 +45,7 @@ func (s *UpgradeTestSuite) TestUpgrade() {
 	ctx := s.Ctx
 	keepers := s.App.AppKeepers
 
-	modulesToRemove := []string{"qtransfer", "qoracle", "qvesting"}
+	modulesToRemove := []string{} // as now the module keys are deprecated and cleaned up post v3 upgrade this fails otherwise
```
Comment (arhamchordia) [inline app/upgrades/v3/upgrades_test.go:48]: Same : We can completely get rid of this v3 upgrade directory here.

File: app/modules.go [hunk 6]
```diff
@@ -245,6 +249,8 @@ func orderBeginBlockers() []string {
 		wasmtypes.ModuleName,
 		ibcwasmtypes.ModuleName,
 		tftypes.ModuleName,
+		marketmaptypes.ModuleName,
```
Comment (arhamchordia) [inline app/modules.go:252]: Wondering if this is the order here? As slinky needs to prioritised then it should be on the top of `begin_blockers`, `end_blockers` and `init_blockers`

File: app/upgrades/v4/upgrades.go (hunk 3 above)
Comment (arhamchordia) [reply app/upgrades/v4/upgrades.go:62]: I would say okay to keep this as gov module.

File: cmd/quasard/cmd/root.go [hunk 7]
```diff
@@ -269,11 +271,7 @@ func txCommand(basicManager module.BasicManager) *cobra.Command {
 		authcmd.GetEncodeCommand(),
 		authcmd.GetDecodeCommand(),
 	)
-
-	// NOTE: this must be registered for now so that submit-legacy-proposal
-	// message (e.g. consumer-addition proposal) can be routed to the its handler and processed correctly.
-	basicManager.AddTxCommands(cmd)
-
+
```
Comment (arhamchordia) [inline cmd/quasard/cmd/root.go:274]: Oh this fixes that multiple commands bug. 🫠

File: go.mod [hunk 8]
```diff
@@ -35,6 +35,7 @@ require (
 	github.com/gorilla/mux v1.8.1
```
Comment (arhamchordia) [inline go.mod:35]: Cannot comment on that line but ibc-go can be updated.
Other bumps can also be a part of this PR or a PR that merges to this branch.

**Review by arhamchordia — CHANGES_REQUESTED** [review]
Rest all looks good. We can merge after resolving comments.

File: app/upgrades/v4/constants.go (hunk 2 above)
Comment (ajansari95) [reply app/upgrades/v4/constants.go:12]: this needs to be setup when we are ready to update

File: app/modules.go (hunk 6 above)
Comment (ajansari95) [reply app/modules.go:252]: so slinky process uses vote extensions to resolve marketmap feeds - the market map module just has the state so matter in ordering

**Review by arhamchordia — APPROVED** [review]
Looks good. Have some WIP things, can be dealt with on in future PRs

**Pull request merged**